app/lib/tool/mirror/mirror_config.dart

```diff
@@ -0,0 +1,181 @@
+    if (normalize) mc.normalize();
+    return mc;
+  }
+
+  factory MirrorConfig.fromYaml(String source, {bool normalize = false}) {
+    final map = json.decode(json.encode(yaml.loadYaml(source)));
+    final mc = MirrorConfig.fromJson(map as Map<String, dynamic>);
+    if (normalize) mc.normalize();
+    return mc;
+  }
+
+  Map<String, dynamic> toJson() => _$MirrorConfigToJson(this);
+
+  void normalize() {
+    packages ??= <Package>[];
+    publishers ??= <Publisher>[];
+    users ??= <User>[];
+
+    for (final publisher in publishers) {
+      publisher.members ??= <Member>[];
+
+      // copy package definitions to [MirrorConfig.packages]
+      if (publisher.packages != null) {
+        publisher.packages.forEach((p) {
+          p.publisher = publisher.name;
+          packages.add(p);
+        });
+        publisher.packages = null;
+      }
+
+      // convert admins to members
+      if (publisher.admins != null) {
+        publisher.members.addAll(
+            publisher.admins.map((e) => Member(email: e, role: 'admin')));
+        publisher.admins = null;
+      }
+
+      // add missing users
+      publisher.members?.forEach((m) => _createUserIfNeeded(m.email));
+    }
+
+    for (final package in packages) {
+      // add missing users
+      package.uploaders?.forEach(_createUserIfNeeded);
+
+      // add missing publishers
+      if (package.publisher != null) {
+        final publisher = publishers
+            .firstWhere((p) => p.name == package.publisher, orElse: () => null);
+        if (publisher == null) {
+          publishers.add(Publisher(name: package.publisher, members: []));
+        }
+      }
+    }
+
+    defaultUser ??= users.first.email;
+    _createUserIfNeeded(defaultUser);
+
+    for (final publisher in publishers) {
+      if (publisher.members.isEmpty) {
+        publisher.members.add(Member(email: defaultUser, role: 'admin'));
+      }
+    }
+
+    for (final package in packages) {
+      if (package.publisher == null) {
+        package.uploaders ??= <String>[];
+        if (package.uploaders.isEmpty) {
+          package.uploaders.add(defaultUser);
+        }
+      }
+    }
+  }
+
+  void _createUserIfNeeded(String email) {
+    final user = users.firstWhere((u) => u.email == email, orElse: () => null);
+    if (user == null) {
+      users.add(User(email: email));
+    }
+  }
+}
+
+@JsonSerializable(explicitToJson: true, includeIfNull: false)
+class Package {
+  final String name;
+  List<String> uploaders;
+  String publisher;
+  List<String> versions;
+
+  Package({
+    this.name,
+    this.uploaders,
+    this.publisher,
+    this.versions,
+  });
+
+  factory Package.fromJson(Map<String, dynamic> json) =>
+      _$PackageFromJson(json);
+
+  Map<String, dynamic> toJson() => _$PackageToJson(this);
+}
+
+@JsonSerializable(explicitToJson: true, includeIfNull: false)
+class Publisher {
+  final String name;
+  List<Member> members;
+  List<Package> packages;
```

Comment:
Maybe we should remove this, because packages can be inserted at the top-level and assigned a `publisher` property, instead of `uploaders`.

I get why we allow for `Package` to have either `uploaders` or `publisher`... In this case, I then propose that we avoid nesting... and only have one way to define it.